### ISSUE TITLE
Feature: custom API key

### ISSUE THREAD

**Issue by noopkat** [ratwithacompiler/OBS-captions-plugin#4]
👋 Hello

Just wanna start off by saying that this is AMAZING. I have wanted native captions in OBS to be decoupled from Windows speech recognizer for a long time because I stream on macOS. Thank you so much for such an awesome plugin!

This plugin could run into a couple of issues which are similar issues to the built-in experimental captions in OBS:

1. It's tightly coupled to Google Cloud service so there's provider lock-in and no stenography service support.
2. There doesn't seem to be anywhere to easily change the API key for the Google Cloud speech service and I'm worried this will end up costing you a lot of money if people use it a lot / you'll have to turn off the service for whatever reason.

Would it be possible to perhaps solve issue 2 to begin with, and allow folks to use their own Google Cloud speech API key?

If issue 1 is too much work to decouple I can always fork this repo as well if that's ok. I thought it would be awesome if we could have a generic caption interleaving plugin like this that folks can have the captions piped to an endpoint or simple local socket so that even human stenographers can be used with livestreams. Just some pipe dreams! 🌈 

Anyway thanks so much again and I hope to hear back from you.
Have a great day 🌻

**Comment by ratwithacompiler** [comment]
Hi!

Regarding custom keys, currently this plugin uses an older HTTP speech recognition API which doesn't accept new API keys anymore but I don't believe there's too much need to worry about service for the existing plugin being turned off anytime soon but Github isn't really the place to discuss that publicly for various reasons ... 😅

I am also working on support for the new Google GRPC cloud speech API, that implementation is not fully finished and stable yet and still private but that would support custom keys if/when that need ever arises in the future.


Regarding a generic caption interface, OBS has Lua and Python scripting support built in (the Python one does require the user to install Python manually and then give the bin path but the Lua one just works out of the box), both of those allow sending captions as well. The main reason I had to go with a native plugin was because that's required to get the actual audio data from OBS sources. 

If your goal is to run a generic http/socket/whatever server in OBS that other tools can connect to and send native OBS captions via it then just doing it via Lua/Python scripting would probably be much easier than forking this and doing it in C++, as long as the external tool can get audio directly via mic or some other way.


I did experiment with making such an interface with via running a GRPC server in OBS a while ago just out of curiosity and got that working but it was not a fun experience tbh. I would only suggest that if you really need the interface to be able to pass out actual audio data to the external tools using it but if you only wanna accept external caption text then scripting would seem like the way to go.

I also saw the VOD of your stream earlier, glad you like my plugin, and also suggested the scripting api approach to Gooseman as well now as a way to get native output for him.

**Comment by noopkat** [comment]
Thank you for such a fast and thorough response. All of this information is so useful for me. The documentation for OBS can be a little cumbersome to find the information I needed so having this explained by someone who is experienced with developing a plugin is all the more helpful as well.

Being able to use the audio source directly from OBS is definitely very convenient so I can see why you chose to write this in C++. I wasn't aware that the Lua and Python scripting supported the captioning functionality so I assumed that was the only option anyway. I could have sworn I checked that earlier this year and didn't see mention of it in the Lua / Python part of the OBS documentation. I'll look into this again! 

I didn't realise you know Gooseman - he is a great person. I love his extension and would love to see it interleaving captions natively into the stream. Perhaps it's something that he and I can work on together we'll see 😄Also thanks for checking out my stream! I was pretty excited by your plugin so I'm glad you got to see that moment.

**Comment by ratwithacompiler** [comment]
No problem!

Yeah the Caption output has only been available to the scripting API as of a few months ago because of a minor oversight in how scripting used to be built, was fixed by a PR from me because that's the way I initially wanted to go for mine too before I learned about the audio access limitation for scripting 😂.

It also used to be limited to only once every two seconds which was fixed by a PR by someone else at about the same time with the added `obs_output_output_caption_text2`. Both PRs were added with OBS 23.2.1 so you are limited to that version or newer (same with native plugins) but most people do update regularly anyway.

Some more things I forgot to mention, the nice thing about the scripting API is also that scripts can be installed via the OBS UI itself in Tools -> Scripts making installing them a lot more user friendly to less tech-savvy people. Scripts also let you declare basic UI forms fields and buttons all within the main scripts window without having to make full Qt interfaces yourself making them pretty powerful still.

Regarding Python, in addition to needing to be installed and the bin path supplied by the user it's also pretty hard to get to work on Mac so Lua is probably the way to go (unfortunately) if you go the scripting route. 

On reread I hope I didn't come off too negatively against forking mine and/or doing it natively in general as it's certainly a good option as well I just heard you mention the thing about this plugin putting you 90% there regarding making a generic interface but while mine does give the basic scaffolding for setting up a plugin and probably some useful bits in general most of it is coupled towards my specific use case so I just didn't see that as being too useful for making an interface. Also just doing anything in C++ just takes SOOOOO LONG (especially setting up builds) but that may just be me though 😬.

Good luck!
